### PR TITLE
RSDK-4393 Upload CSI module to registry and set up release workflow

### PR DESCRIPTION
**Pull request by bazile-clyde**
The only thing left is to create a release tag, since this is only uploaded on tags that match the version `v[0-9]+.[0-9]+.[0-9]+`.

@abe-winter if you can give this a look and make sure I'm not doing anything obviously wrong. I followed the instructions in [upload-module](https://github.com/viamrobotics/upload-module#upload-module-action) as closely as I could. It looks like lots of modules also add 
```
   permissions:
      packages: write
```
I planned on trying without it and if it's needed I'll update the yaml file here and submit a PR to the instructions to clarify.

@seanavery if you can take a look at the build stuff and let me know if anything's off I'd appreciate it. I used the image `ghcr.io/seanavery/viam-cpp-jetson:0.0.1` but I'm unsure if that already has the dependencies installed on it, so I install them during the run just to be safe.

### CODE REVIEW

**Comment by bazile-clyde** [comment]
@seanavery I do not have permission to add the `viam_key_id` and `viam_key_value` secrets to this repo. Could you either give me permission or add them?

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  push:
+    tags:
+      - 'v[0-9]+.[0-9]+.[0-9]+'
+
+jobs:
+  publish:
+    runs-on: [buildjet-2vcpu-ubuntu-2204-arm]
+    container:
+      image: ghcr.io/seanavery/viam-cpp-jetson:0.0.1
+      options: --platform linux/arm64
```
Comment (abe-winter) [inline .github/workflows/deploy.yml:11]: you probably don't need the `--platform` switch unless you're using QEMU to build a platform that is non-native to the runner

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+on:
+  push:
+    tags:
+      - 'v[0-9]+.[0-9]+.[0-9]+'
```
Comment (abe-winter) [inline .github/workflows/deploy.yml:4]: afaik we don't allow v1.2.3 type versions in the registry (must be 1.2.3 format). We may be allowing them now, but maybe change this to make the `v` optional
```suggestion
      - 'v?[0-9]+.[0-9]+.[0-9]+'
```

**Review by abe-winter — APPROVED** [review]
lgtm!

re `packages: write`, you only need it if you're writing a ghcr image

**Comment by seanavery** [comment]
@bazile-clyde Added you as admin to the repository. Should be able to add deploy keys now.

**Comment by seanavery** [comment]
> @seanavery if you can take a look at the build stuff and let me know if anything's off I'd appreciate it. I used the image ghcr.io/seanavery/viam-cpp-jetson:0.0.1 but I'm unsure if that already has the dependencies installed on it, so I install them during the run just to be safe.

Base level image is tagged to my personal profile. (Still need to upload to viamrobotics - will create a ticket for this).

All all deps for viam-cpp-sdk and making appimage are installed in the base container. `make dep` installs all the module-specific libraries (GST). So yes, good to run in the pipeline before building.

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    - name: Build viam-csi binary
+      run: make build
+
+    - name: Build viam-csi appimage
+      run: make package
+
+    - name: Copy binary and appimage to bin folder and archive
+      run: make bin && tar czf module.tar.gz ./bin/*.AppImage
+
```
Comment (seanavery) [inline .github/workflows/deploy.yml:30]: We may want to run the sanity test before deploying to make sure that module is working in fresh L4T instance.

```
      - name: Run appimage in fresh L4T instance
        run: |
          make image-test
          make test-package
```

File: .github/workflows/deploy.yml (hunk 1 above)
Comment (seanavery) [reply .github/workflows/deploy.yml:11]: Ok - then we probably want to update the main.yaml as well.

Also, no need to use `matrix` in main...

**Review by seanavery — COMMENTED** [review]
@bazile-clyde this is looking good! A couple of questions/comments.

- We should update README install instructions to point to module registry.
- Can we reuse steps between yaml files (main and deploy)?
- Should we be deploying "latest" to module registry for all main pushes?
- Curious on how you are testing this pipeline out on releases.
- Maybe we should make sure uploaded module is working on fresh jetson machine before merging.

**Comment by tessavitabile** [comment]
Is this ready to merge? We would like to get all modules in the registry.

**Comment by bazile-clyde** [comment]
@seanavery 

> * We should update README install instructions to point to module registry.
> 

I don't think so. We expect users to discover modules via the registry, and the registry links to both docs on how to automatically deploy modules and this GitHub repo. We haven't been linking to our docs or duplicating instructions in our repos because the docs change often, and want a single source of truth.

> * Can we reuse/abstract common steps between yaml files (main and deploy)?
> 

This would be nice but I don't think it's possible. AFAIK there's no way to _include_ or _import_ yaml files and putting the instructions in a single script would make the steps for uploading the module more opaque in GitHub CI. If a single step failed, we'd want to know exactly which step failed.

> * Should we be deploying "latest" to module registry for all main pushes?
> 

No. We only want to deploy on release versions. _latest_ is always unstable. For similar reasons, this is why our app links a _stable_ version of the RDK and not _latest_.

> * Curious on how you are testing this pipeline out on releases.
> 

I haven't yet and can't until we actually release this repo since the script is only executed _after_ a new release is published. I will do it after this is set up and then manually test everything. It'd be nice to test this before uploading, but there isn't a good way to do an e2e test since there's no concept of a staging registry, AFAIK. I'm not too concerned about it, though, since nothing gets deployed if this doesn't work properly, and no users would be affected.

File: .github/workflows/deploy.yml (hunk 3 above)
Comment (bazile-clyde) [reply .github/workflows/deploy.yml]: Good call! Done.

File: .github/workflows/deploy.yml (hunk 1 above)
Comment (bazile-clyde) [reply .github/workflows/deploy.yml]: Oh, good catches! I removed `--platform` from this file and `main.yml`. I didn't remove the `matrix` though. Seems like a good thing to leave in to future-proof the setup and I wouldn't want to make any potentially breaking changes in this PR.

Comment (bazile-clyde) [reply .github/workflows/deploy.yml:11]: Actually, removing `--platform` breaks the docker configuration... [link](https://github.com/viamrobotics/csi-camera/actions/runs/6226719932/job/16899817266). Leaving it in.

**Comment by seanavery** [comment]
>  don't think so. We expect users to discover modules via the registry, and the registry links to both docs on how to automatically deploy modules and this GitHub repo. We haven't been linking to our docs or duplicating instructions in our repos because the docs change often, and want a single source of truth.

Ok - yes that would be just circular referencing.

> No. We only want to deploy on release versions. latest is always unstable. For similar reasons, this is why our app links a stable version of the RDK and not latest.

Sounds good about only updating the registry on proper releases.

**Review by seanavery — APPROVED** [review]
LGTM!

**Pull request merged**